### PR TITLE
Fix an issue with loading bundles with multi-byte character

### PR DESCRIPTION
**Pull request by wu-hui**
Both `BundleReader` and the testing `BundlerBuilder` should be counting UTF-8 bytes, instead of Java `char` count.

This is a fix for: https://github.com/firebase/firebase-android-sdk/issues/2805

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 1]
```diff
@@ -153,29 +152,50 @@ private int indexOfOpenBracket() {
     }
   }
 
+  private class ReadJsonResult {
```
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:155]: drive-by comment: this class should be `static`, and probably `final`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 2]
```diff
@@ -153,29 +152,50 @@ private int indexOfOpenBracket() {
...
       if (buffer.remaining() == 0 && !pullMoreData()) {
         throw abort("Reached the end of bundle when more data was expected.");
       }
 
       int read = Math.min(remaining, buffer.remaining());
-      json.append(buffer, 0, read);
-      buffer.position(buffer.position() + read);
+      byte[] bytes = new byte[read];
+      buffer.get(bytes);
+      json.append(new String(bytes, charset));
```
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:192]: drive-by comment: This is dangerous because `bytes` could end in the middle of a multi-byte character. You could possibly use a class like `java.nio.charset.CharsetDecoder` to handle this properly. 

https://docs.oracle.com/javase/7/docs/api/java/nio/charset/CharsetEncoder.html

**Review by dconeybe — CHANGES_REQUESTED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 1 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 2 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Done.

Comment (dconeybe) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: The change doesn't fix the issue. For example, consider the situation where the last byte of `bytes` specified to `new String(bytes, charset)` is the first byte of a multi-byte character, and the second byte of that character would be the first byte of the `bytes` in the next iteration; this would probably either fail decoding with an exception or would produce garbage.

**Review by dconeybe — CHANGES_REQUESTED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 2 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: This is the bytes the bundle tells the SDK that it is a UTF8 string. If what you described happens, it means the bundle is invalid, and an exception is exactly what we want here.

Comment (dconeybe) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: IIUC the code reads the bundle's bytes in "chunks". Wouldn't there be arbitrary breaks in the "chunks" such that a multibyte character could span two chunks?

Comment (dconeybe) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: But now, what if `bytesToRead` > `BUFFER_CAPACITY`? IIUC, with the updated code, if the buffer fills up then `pullMoreData()` will return `false` and `readJson()` will incorrectly throw `IllegalArgumentException`.

I have another idea. What if you go back to the original implementation but change `StringBuilder` to `ByteArrayOutputStream`. You can then accumulate the bytes in the `ByteArrayOutputStream` and, once all bytes are read, you can decode the entire byte sequence, avoiding the multi-byte-character-spanning-a-chunk issue.

Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Yeah, that is better.

File: firebase-firestore/CHANGELOG.md [hunk 3]
```diff
@@ -1,10 +1,18 @@
-# Unreleased
-- [changed] Increases the aggressiveness of network retires when an app's
+# 23.0.3
+- [fixed] Fixed an issue when loading a data bundle with
+  multi-bytes unicode character leads to failures.
```
Comment (dconeybe) [inline firebase-firestore/CHANGELOG.md:3]: nit: "multi-bytes unicode character" -> "multi-byte Unicode characters". The "U" in "Unicode" should be uppercase and the plural needs to be tweaked.

Comment (wu-hui) [reply firebase-firestore/CHANGELOG.md]: Done.

**Review by dconeybe — CHANGES_REQUESTED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 4]
```diff
@@ -35,17 +35,17 @@
   protected static final int BUFFER_CAPACITY = 1024;
 
   private final BundleSerializer serializer;
-  private final InputStreamReader dataReader;
+  private final InputStream bundleInputStream;
   private final Charset charset = Charset.forName("UTF-8");
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:39]: Unfortunately, this was only added with API level 19.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 5]
```diff
@@ -101,8 +101,9 @@ private BundleElement readNextElement() throws IOException, JSONException {
       return null;
     }
 
-    String json = readJsonString(Integer.parseInt(lengthPrefix));
-    bytesRead += lengthPrefix.length() + json.getBytes(charset).length;
+    int jsonStringByteCount = Integer.parseInt(lengthPrefix);
+    String json = readJsonString(jsonStringByteCount);
+    bytesRead += lengthPrefix.getBytes(charset).length + jsonStringByteCount;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:106]: This should just be the lengthPrefix, which should not have UTF-8 as it is just a number.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 6]
```diff
@@ -133,9 +134,9 @@ private BundleElement readNextElement() throws IOException, JSONException {
       throw abort("Reached the end of bundle when a length string is expected.");
     }
 
-    char[] c = new char[nextOpenBracket];
+    byte[] c = new byte[nextOpenBracket];
     buffer.get(c);
-    return new String(c);
+    return charset.decode(ByteBuffer.wrap(c)).toString();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:139]: Note that `indexOfOpenBracket` returns the char-index of the first bracket. Using a byte buffer here is not correct. but likely won't make a difference since this part of the input should be ASCII.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 7]
```diff
@@ -157,25 +158,30 @@ private int indexOfOpenBracket() {
...
     while (remaining > 0) {
       if (buffer.remaining() == 0 && !pullMoreData()) {
         throw abort("Reached the end of bundle when more data was expected.");
       }
 
+      // `read` is the number of bytes guaranteed to exist in `this.buffer` after the above
+      // call to `pullMoreData`. Copy them to `jsonBytes` and advance `this.buffer`'s position.
       int read = Math.min(remaining, buffer.remaining());
-      json.append(buffer, 0, read);
+      jsonBytes.write(buffer.array(), buffer.position(), read);
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:178]: More over, `buffer` is a CharBuffer. CharBuffer deals with UTF-8 direcltly:

```
            String unicode = "\uD83D\uDE0A";
            byte[] b = unicode.getBytes();
            CharBuffer buffer = CharBuffer.allocate(10);
            new InputStreamReader(new ByteArrayInputStream(b)).read(buffer);
            StringBuilder result = new StringBuilder();
            buffer.flip();
            System.out.println(buffer.remaining());
            result.append(buffer);
            System.out.println(result);
```

Prints:
```
2
😊
```

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java [hunk 8]
```diff
@@ -238,18 +239,22 @@ private void verifyQueryResults() throws ExecutionException, InterruptedExceptio
...
     QuerySnapshot limitSnapshot = Tasks.await(limitQuery.get(Source.CACHE));
-    assertEquals(asList(map("bar", 2L, "k", "b")), querySnapshotToValues(limitSnapshot));
+    assertEquals(asList(map("bar", 2L, "k", "\uD83D\uDE0A")), querySnapshotToValues(limitSnapshot));
 
     Query limitToLastQuery = Tasks.await(db.getNamedQuery("limit-to-last"));
     QuerySnapshot limitToLastSnapshot = Tasks.await(limitToLastQuery.get(Source.CACHE));
     assertEquals(asList(map("bar", 1L, "k", "a")), querySnapshotToValues(limitToLastSnapshot));
   }
 
+  private int getUTF8BytesCount(String s) {
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: It's only supported after KitKat, which I think is why we use string name here.

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java [hunk 9]
```diff
@@ -71,7 +72,7 @@
             + "{\"seconds\":1000,\"nanos\":9999},\"exists\":true}}",
         "{\"document\":{\"name\":\"projects/{projectId}/databases/(default)/documents/coll-1/b\","
             + "\"createTime\":{\"seconds\":1,\"nanos\":9},\"updateTime\":{\"seconds\":1,"
-            + "\"nanos\":9},\"fields\":{\"k\":{\"stringValue\":\"b\"},\"bar\":"
+            + "\"nanos\":9},\"fields\":{\"k\":{\"stringValue\":\"\uD83D\uDE0A\"},\"bar\":"
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java:75]: Note that this test passes against `master` as well.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 10]
```diff
@@ -185,9 +186,12 @@ private String readJsonString(int length) throws IOException {
    */
   private boolean pullMoreData() throws IOException {
     buffer.compact();
-    int read = dataReader.read(buffer);
+    int bytesToRead = Math.min(bundleInputStream.available(), buffer.remaining());
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:194]: We need a test that hits this issue.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 11]
```diff
@@ -185,9 +186,12 @@ private String readJsonString(int length) throws IOException {
...
+    byte[] bytes = new byte[bytesToRead];
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:195]: s/Consider reading/Please read

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: This is only for the length string, which is typically 3 to 4 bytes. IMO, it's not worth for the optimization.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: See above.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Thanks!

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:39]: Ahh nevermind. I'll resolve this comment.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:203]: Similarly, I feel better to avoid performance optimization without proof. The end result is some byte counting, which is bad for readability.

The other place we do this is acceptable, because there is no good alternative..

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:178]: FYI `buffer` was changed from `CharBuffer` to `ByteBuffer` in one of the recent commits.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:178]: Got it, but the existing code uses it throughout and creates the initial input stream using UTF-8. It should work (and the test passes), so I am a bit concerned that the bug is somewhere else.

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java (hunk 9 above)
Comment (wu-hui) [reply firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java:76]: Because the bundles built here also count chars, not bytes.

Comment (wu-hui) [reply firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java:76]: See other change in this test, it uses byte count, which fails the tests.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 12]
```diff
@@ -191,12 +192,23 @@ private String readJsonString(int bytesToRead) throws IOException {
...
+    // forward.
+    if (available == 0) {
+      bytesToRead = buffer.remaining();
+    }
     byte[] bytes = new byte[bytesToRead];
     int bytesRead = bundleInputStream.read(bytes);
-    buffer.put(bytes);
+    boolean readSuccess = bytesRead > 0;
+    if (readSuccess) {
+      buffer.put(bytes, 0, bytesRead);
```
Comment (dconeybe) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:207]: Could you make the call to `buffer.put(bytes, 0, bytesRead)` unconditional? Or is specifying `bytesRead==0` going to cause problems?

**Review by dconeybe — APPROVED** [review]
I approve, but please ensure that Sebastian's comments are resolved.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Yeah, the tests pass because the test bundle builder also count characters, not bytes.

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java (hunk 9 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/androidTest/java/com/google/firebase/firestore/BundleTest.java:76]: Sorry, I just saw this flaw in my logic. So the bug is that we try to read more data than available.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: To reply to Sebastian: I am not sure how to do this without mocking, `available()` is stated to be an estimate, which is pretty accurate for most implementations.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:203]: You are doubling the amount of memory consumed here. Gil pushed back heavily against this in the original PR.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 12 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:207]: `bytesRead` can be -1, that is the problem.

Comment (dconeybe) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:207]: Ahh I see. Makes sense.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 13]
```diff
@@ -34,18 +34,19 @@
...
   @Nullable BundleMetadata metadata;
-  private CharBuffer buffer;
+  private ByteBuffer buffer;
   long bytesRead;
 
   public BundleReader(BundleSerializer serializer, InputStream data) {
     this.serializer = serializer;
-    dataReader = new InputStreamReader(data, charset);
-    buffer = CharBuffer.allocate(BUFFER_CAPACITY);
+    bundleInputStream = data;
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:48]: Can you rename the argument to `bundleInputStream` as well? We could also drop the prefix similar to `serializer`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: I still think this is premature..but done anyways.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 13 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 14]
```diff
@@ -133,9 +135,9 @@ private BundleElement readNextElement() throws IOException, JSONException {
       throw abort("Reached the end of bundle when a length string is expected.");
     }
 
-    char[] c = new char[nextOpenBracket];
+    byte[] c = new byte[nextOpenBracket];
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:138]: Nit: s/c/b

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java [hunk 15]
```diff
@@ -185,9 +192,24 @@ private String readJsonString(int length) throws IOException {
...
     buffer.compact();
-    int read = dataReader.read(buffer);
+
+    int available = bundleInputStream.available();
+    int bytesToRead = Math.min(available, buffer.remaining());
+    // `available` is an estimation, we still try to read if the estimation is 0, to move things
+    // forward.
+    if (available == 0) {
+      bytesToRead = buffer.remaining();
+    }
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java:202]: Can we always read buffer.remaining?

**Review by schmidt-sebastian — APPROVED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 14 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java (hunk 15 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java]: Yes, since we are reading into the buffer directly now!

**Review by schmidt-sebastian — APPROVED** [review]
LGTM
